src/Bitmap_cubical_complex/include/gudhi/Persistence_on_rectangle.h

```diff
@@ -0,0 +1,596 @@
+    // parallel with the primal if we were motivated...
+    tbb::parallel_sort(edges.begin(), edges.end());
+#else
+    std::sort(edges.begin(), edges.end());
+#endif
+#ifdef DEBUG_TRACES
+    std::clog << "edges\n";
+    for(auto&e : edges){ std::clog << e.v1 << '\t' << e.v2 << '\t' << e.filt() << '\n'; }
+#endif
+  }
+
+  template<class Out>
+  void primal(Out&&out){
+    auto it = std::remove_if(edges.begin(), edges.end(), [&](Edge& e) {
+        assert(e.v1 < e.v2);
+        Index a = ds_find_set_vertex(e.v1);
+        Index b = ds_find_set_vertex(e.v2);
+        if (a == b) return false;
+        if (data_vertex(b) < data_vertex(a)) std::swap(a, b);
+        ds_parent_vertex(b) = a;
+        out(data_vertex(b).out(), e.f.out());
+        return true;
+    });
+    edges.erase(it, edges.end());
+    global_min = data_vertex(ds_find_set_vertex(0)).out();
+  }
+
+  // In the dual, squares behave like vertices, and edges are rotated 90° around their middle.
+  // To handle boundaries correctly, we imagine a single exterior cell with filtration +inf.
+  template<class Out>
+  void dual(Out&&out){
+    for (auto e : boost::adaptors::reverse(edges)) {
+      dualize_edge(e);
+      Index a = ds_find_set_square(e.v1);
+      Index b = ds_find_set_square(e.v2);
+      GUDHI_CHECK(a != b, std::logic_error("Bug in Gudhi"));
+      // This is more robust in case the input contains inf? I used to set the filtration of 0 to inf.
+      if (b == 0 || (a != 0 && input(a) < input(b))) std::swap(a, b);
+      ds_parent_square(b) = a;
+      if constexpr (output_index)
+        out(e.f.out(), b);
+      else
+        out(e.f.out(), input(b));
+    }
+  }
+};
+// Ideas for improvement:
+// * for large hard (many intervals) inputs, primal/dual dominate the running time because of the random reads in find_set and input(a/b). The input load would be cheaper if we stored it with parents, but then find_set would be slower.
+// * to increase memory locality, maybe pairing, which is currently arbitrary, could use some heuristic to favor some pairs over others.
+// * try to loosen tight dependency chains, load values several instructions before they are needed. Performing 2 find_set in lock step surprisingly doesn't help.
+// * To handle very big instances, we could remove data_v_ and recompute it on demand as the min of the inputs of i, i+1, i+dy and i+dy+1. On hard instances, it wouldn't save that much memory (and it is a bit slower). On easy instances, if we also remove the calls to reserve(), the saving is less negligible, but we still have ds_parent_*_ that take about as much space as the input. We could, on a subarray, fill a dense ds_parent, then reduce it and export only the critical vertices and boundary to some sparse datastructure, but it doesn't seem worth the trouble for now.
+
+/**
+ * @private
+ * Compute the persistence diagram of a function on a 2d cubical complex, defined as a lower-star filtration of the values at the top-dimensional cells.
+ *
```

Comment:
Why not exposing the documentation just like for persistence on a line ?

```suggestion
 *
 * \ingroup persistent_cohomology
 *
```

Reply:
Sorry, I missed your point `For now only available through the Python CubicalPersistence, so we have more time to think about the exact C++ interface we want to commit to`. Forget this comment.